src/java.base/share/classes/jdk/internal/classfile/impl/StackMapGenerator.java

```diff
@@ -1066,12 +1066,12 @@ void setLocalsFromArg(String name, MethodTypeDesc methodDesc, boolean isStatic,
                     locals[localsSize + 1] = Type.DOUBLE2_TYPE;
                     localsSize += 2;
                 } else {
-                    if (desc == CD_int || desc == CD_boolean || desc == CD_byte || desc == CD_char || desc == CD_short) {
-                        type = Type.INTEGER_TYPE;
+                    if (desc instanceof ReferenceClassDescImpl) {
+                        type = Type.referenceType(desc);
```

Comment:
Not for this PR, but `Type.referenceType(ClassDesc desc)` is always just creating a new `Type` - perhaps it would be profitable to do some quick checks such as `if (desc == CD_Object) return TYPE_OBJECT;` in that method?